### PR TITLE
[Bugfix][Core] Schedule async chunks atomically

### PR DESCRIPTION
**Pull request by dongteng**
## Purpose

Related to vllm-project/vllm#44933.

Prevent connector-delivered async chunks from being clipped to the
remaining scheduling budget. In our local trace, 57 chunks of 1168 tokens
hit B=65536: 56 were scheduled whole and the last was clipped to 128
tokens (`num_scheduled_tokens` shows 56 * 1168 + 128 = 65536). The
clipped codebook-major slice still carried metadata describing the
complete first chunk, so the ICL first-chunk invariant check raised and
killed the whole stage engine.

The change defers async chunks intact in both the waiting and running
scheduling paths when the remaining budget is insufficient, and rejects
ready chunks larger than the full step budget through the standard
finish path plus an explicit ERROR output (emitted even by zero-token
steps). Non-chunk fast-path behavior is kept unchanged. Decoder
validation remains enabled.

## Test Plan

New unit tests in tests/core/sched/test_generation_scheduler_atomic_chunk.py
cover: waiting and running chunks under both chunked-prefill settings,
exact-fit boundaries, non-chunk and sender-only adapters (legacy behavior
preserved), oversized rejection (single, batched, boundary budgets),
zero-token step error emission, error-exactly-once across schedules, and
non-zero client-index routing. Full scheduler-suite regression included.

Reproduction setup (identical for the Before/After runs below): two-stage
Qwen3-TTS pipeline (talker + code2wav) with async_chunk enabled and a
precomputed ICL voice profile (310-frame reference audio, last 72 frames
shipped as the first-chunk prefix); stage-1 (code2wav)
`max_num_batched_tokens=65536`, `max_num_seqs=128`; connector chunking
`codec_chunk_frames=25`, `codec_left_context_frames=72`,
`initial_codec_chunk_frames=1`, so an ICL first chunk is
(72+1)*16 = 1168 tokens. Load: `vllm bench serve --omni --backend
openai-audio-speech` with a zh seed-tts text dataset (~74s of audio per
request), concurrency 80, 300 requests.

**vLLM Version:** v0.28.0 (editable install, paired with the vllm-omni
commit below)

**vLLM-Omni Commit:** eb11446b (branch base of
fix/atomic-async-chunk-scheduling)

## Test Result

- Before (unpatched, setup above): reproduced deterministically -- the
  stage engine died on "ValueError: Qwen3-TTS async_chunk first ICL chunk
  must include its declared reference prefix" (qwen3_tts_code2wav.py:375,
  same site as our 2026-09-10 production incident) within one second of
  the burst; 0/300 requests completed, /health 503 afterwards
- Focused unit tests: python -m pytest
  tests/core/sched/test_generation_scheduler_atomic_chunk.py -q
  -> 20 passed
- Scheduler regression: python -m pytest tests/core/sched/ -m cpu -q
  -> 182 passed
- Lint/type: ruff check + ruff format --check (0.14.10, both files) and
  the repo's mypy-3.10 pre-commit hook -> all clean
- GPU end-to-end (patched, same setup as the Before run): 300 completed /
  0 failed in 225s (peak concurrency 107); zero "first ICL chunk" and
  zero "not divisible" log signatures during traffic; /health 200 after
  the run; manual short synthesis OK.

Deterministic unit-test behavior at B=65536: 56 whole 1168-token chunks
scheduled (65408 total), the 57th preserved intact and deferred to a
later step (not promised to be globally first in the next step).

### CODE REVIEW

**Comment by dongteng** [comment]
Trigger in one line (full repro in the description): when async chunks
arriving within one step fill the token budget, the last chunk is clipped
to the remaining budget while its metadata still describes the full chunk
— e.g. 57 ready first chunks of (72+1)*16 = 1168 tokens vs a 65536 budget
→ the 57th is cut to 128 tokens (8 frames < the 72-frame reference prefix).
- **No behavior change for non-chunk requests**: the waiting/running fast
  paths only defer when `_requires_atomic_chunk()` holds; non-chunk inputs
  keep the exact previous `min(required_tokens, token_budget)` behavior.
  Covered by the parametrized non-chunk tests.
- **Deferral cannot starve the loop**: a chunk larger than the full step
  budget (`max_num_scheduled_tokens`) is rejected up front via
  `finish_requests` with an explicit ERROR output, so scheduling always
  makes progress. Covered by the oversized single/batched/boundary tests.
- **The ERROR output is emitted exactly once per rejected request**, even
  when the rejecting step schedules zero tokens (queued in the scheduler,
  flushed by `update_from_output`). Covered by the error-exactly-once and
  client-index routing tests.
- **Local gates**: 20 new unit tests pass; the existing scheduler suite
  (`tests/core/sched/ -m cpu`, 182 tests) passes; ruff + mypy clean via
  pre-commit.
- **GPU end-to-end on the reproduction setup**: before the patch the crash
  reproduces deterministically within ~1 s under load (0/300 done,
  `ValueError` at `qwen3_tts_code2wav.py:375`, /health 503); after the
  patch 300/300 complete in 225 s with /health 200 and no error signatures.

File: vllm_omni/core/sched/omni_generation_scheduler.py [hunk 1]
```diff
@@ -39,6 +39,63 @@ def __init__(self, *args, **kwargs):
...
+            return
+
+        # Batch cleanup avoids repeatedly scanning queues for many bad requests.
+        # Keep the existing finish path: it reconciles queue status and handles
+        # connector cleanup, finished IDs and deferred KV-block release.
+        finished = self.finish_requests(set(rejected_reasons), RequestStatus.FINISHED_ERROR)
+        for finished_request in finished:
+            # This is a terminal request error, not a resumable segment boundary.
+            finished_request.resumable = False
+            error_output = self._make_omni_engine_output(
```
Comment (hsliuustc0106) [inline vllm_omni/core/sched/omni_generation_scheduler.py:91]: P3 (question): for a streaming-enabled request this stop_reason never reaches the client — the orchestrator's raw-terminal path (`_apply_raw_terminal_stage_finish` fires on finish_reason set + is_segment_finished=False) routes to `_finish_raw_terminal_requests`, which emits an empty RequestOutput with finish_reason="stop". So the reject detail lands in engine logs only and the client sees a silent empty success. Pre-existing shape for resumable=False terminations, but this is the first path that queues an ERROR output for a never-scheduled request — is the empty-success downgrade intended, or should the final stage's processed terminal carry the error?

File: vllm_omni/core/sched/omni_generation_scheduler.py [hunk 2]
```diff
@@ -39,6 +39,63 @@ def __init__(self, *args, **kwargs):
         self._init_omni_io_scheduling_state()
         self._retains_state_across_chunks = bool(getattr(model_config, "retains_state_across_chunks", False))
         self._pending_finish_reqs: list[Request] = []
+        self._pending_chunk_error_outputs: list[tuple[int, EngineCoreOutput]] = []
+
+    def _requires_atomic_chunk(self) -> bool:
+        adapter = self.chunk_transfer_adapter
+        return adapter is not None and adapter.receives_chunks
+
+    def _reject_oversized_ready_chunks(self) -> None:
```
Comment (hsliuustc0106) [inline vllm_omni/core/sched/omni_generation_scheduler.py:48]: P3 (docs): the async-chunk scheduling contract this PR establishes — chunks are scheduled atomically (deferred intact, never split) and oversized ready chunks are rejected terminally — isn't recorded in docs/design/feature/async_chunk.md (§"Schedulers" still describes the process/restore flow without the atomicity guarantee). Worth a short paragraph there.

File: tests/core/sched/test_generation_scheduler_atomic_chunk.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+import vllm_omni  # noqa: F401 - import for side effects (patch vLLM)
+from vllm.sampling_params import SamplingParams
+from vllm.v1.core.sched.interface import PauseState
+from vllm.v1.core.sched.request_queue import SchedulingPolicy, create_request_queue
+from vllm.v1.engine import FinishReason
+from vllm.v1.request import Request, RequestStatus
+from vllm_omni.core.sched.omni_generation_scheduler import OmniGenerationScheduler
+# isort: on
+
+pytestmark = [pytest.mark.core_model, pytest.mark.cpu]
```
Comment (hsliuustc0106) [inline tests/core/sched/test_generation_scheduler_atomic_chunk.py:42]: P3 (CI): these cpu-marked tests haven't executed in any CI lane on this PR (no test labels applied; GHA here is DCO/build/pre-commit only), so the 20-passed and 182-passed results are author-run only. Worth a CPU lane run before merge — the waiting-loop split reproduction (chunked-prefill=True param) is the one that fails on the base commit.

**Comment by amy-why-3459** [comment]
Reviewed commit `bd2a5aea8640`. The atomic scheduling change looks well scoped: deferring a whole connector-delivered chunk addresses the metadata/input mismatch, and the existing cleanup path is reused without changing non-chunk scheduling behavior.

### [P2] Preserve oversized-chunk errors through the raw-terminal fallback

The new rejection path emits `FinishReason.ERROR` with the budget-exceeded reason. However, when a streaming request reaches the orchestrator's raw-terminal fallback without a processed terminal output, `_finish_raw_terminal_requests()` constructs an empty completion with `finish_reason="stop"` and `stop_reason=None`. The rejection can therefore lose its error status and diagnostic message at this boundary.

This agrees with the existing review comment. Using the existing orchestrator test harness, I injected an ERROR terminal with the oversized-chunk reason and reproduced a final `stop / None` result on the no-processed-terminal path. This verifies that fallback branch; it is not a real HTTP end-to-end reproduction and does not establish that every oversized rejection takes this path.

Please preserve `FinishReason.ERROR` and the rejection reason through the raw-terminal fallback, rather than converting them into an empty successful completion. Add a regression test covering an oversized first chunk through the output-processing/orchestrator boundary, and assert that the final result retains the error and completes exactly once.

### Non-blocking suggestions

- **Exercise an actual zero-token engine step.** The current tests manually call `update_from_output()`. They demonstrate that queued errors drain correctly, but do not establish that an engine step which only rejects requests and schedules no tokens actually delivers the terminal output. A focused engine-level test would close this gap.
- **Document the scheduling contract.** Please record in the async-chunk design documentation that chunks are deferred intact when the remaining budget is insufficient, chunks exceeding the full step budget are rejected terminally, and non-chunk behavior is unchanged.

Validation: all 20 new tests passed. With vLLM 0.28.0, the scheduler suite produced 179 passes and 3 failures; all three logging-capture failures also reproduced on the pre-change base. The additional orchestrator fallback probe passed. Ruff check and format check passed. I did not run GPU or HTTP end-to-end validation for this review.

File: vllm_omni/core/sched/omni_generation_scheduler.py (hunk 1 above)
Comment (dongteng) [reply vllm_omni/core/sched/omni_generation_scheduler.py:91]: Thanks, you're right — the empty-success downgrade was not intended. The scheduler-level tests verified the terminal EngineCoreOutput, but missed the streaming orchestrator path: the raw-terminal accumulator retained only the request ID, and `_finish_raw_terminal_requests` consequently replaced `FinishReason.ERROR` and its `stop_reason` with an empty successful STOP output.

I updated the raw-terminal handling to preserve the original terminal output. A raw `FinishReason.ERROR` from any stage now takes the request-scoped error path before the normal all-final-stage empty-output fallback, keeps the exact rejection detail, and aborts the remaining stage work. Non-final-stage ERROR terminals are routed the same way, so a mid-pipeline chunk rejection cannot leave the request hanging. Normal non-error terminals retain the existing behavior. I also added orchestrator regressions that drive the raw ERROR through the output-processing/orchestrator boundary and verify one terminal error plus request cleanup.

Done in 5197cef2.

File: vllm_omni/core/sched/omni_generation_scheduler.py (hunk 2 above)
Comment (dongteng) [reply vllm_omni/core/sched/omni_generation_scheduler.py:48]: Added under `Architecture > Key Components > Schedulers` in `docs/design/feature/async_chunk.md`: async chunks are indivisible scheduling units in `OmniGenerationScheduler`, oversized chunks are rejected with `FINISHED_ERROR` plus an ERROR output, and non-chunk inputs keep the splittable behavior. Done in a47383b0.

**Comment by dongteng** [comment]
@amy-why-3459 Thanks for the validation and the precise test boundary. I fixed the raw-terminal downgrade in 5197cef2. A raw ERROR terminal is now routed before the normal empty-terminal fallback, with the original `stop_reason` preserved, and the remaining stage work is aborted during request cleanup. Non-final-stage ERROR terminals are routed the same way, so a mid-pipeline chunk rejection cannot leave the request hanging.

I extended `tests/engine/test_orchestrator.py` with regressions that inject the oversized-chunk ERROR as a raw engine output, drive it through the output-processing/orchestrator path, and verify the exact error, exactly one terminal result, and cleanup. I also covered the multi-final-stage case and kept normal non-error terminals unchanged.

Local results: `tests/core/sched/test_generation_scheduler_atomic_chunk.py` → 20 passed; `tests/engine/test_orchestrator.py` → 62 passed; `tests/core/sched/ -m cpu` → 182 passed; `tests/engine/ -m cpu` → 591 passed, 10 failed (all 10 in `test_arg_utils.py`, reproducing identically on the pre-PR base v0.28.0).
The scheduling contract is documented in a47383b0.

**Comment by dongteng** [comment]
Hi @hsliuustc0106 and @Gaohan123, gentle ping on this PR. I've resolved the earlier merge conflicts, and the other checks are passing, but the required `buildkite/vllm-omni` check is still showing “Expected — Waiting for status to be reported.” Could you help check whether the Buildkite run was triggered? I'd also appreciate a review when you have time. I'm happy to address any feedback or update the branch again if needed. Thanks!
